README.md
<!-- start badges -->
![GitHub release](https://img.shields.io/github/release/pablomenino/fedora-install-script)
![GitHub license](https://img.shields.io/github/license/pablomenino/fedora-install-script)
![GitHub last commit](https://img.shields.io/github/last-commit/pablomenino/fedora-install-script)
![GitHub repo size](https://img.shields.io/github/repo-size/pablomenino/fedora-install-script)
![Contributors](https://img.shields.io/github/contributors-anon/pablomenino/fedora-install-script)
![GitHub followers](https://img.shields.io/github/followers/pablomenino?label=Follow)
![Twitter Follow](https://img.shields.io/twitter/follow/pmenino)
<!-- end badges -->

<!-- start description -->
<h1>Welcome to fedora-install-script 👋</h1>
<p>
    <a href="https://lab.mfwlab.com/lab/fedora-install-script/" id="homepage" rel="nofollow">
        <img align="right" height="128" id="icon" src="logo.svg" width="128"/>
    </a>
</p>
<h2>🏠 <a href="https://lab.mfwlab.com/lab/fedora-install-script/" id="homepage">Homepage</a></h2>
<p>
    fedora-install-script is a Installation Script, that can filter hostname to select the node destination and have a pre-configured list of selected packages to install, and/or filter by category of packages to install.
</p>
<!-- end description -->

## Table of contents

* [How it works](#how-it-works)
* [How to Use](#how-to-use)
* [Support me](#support-me)
* [Third party](#third-party)

## <a name="how-it-works"> How it works

I have several VM whit Fedora, this scripts are to automatically install a pre configured list of packages on this new installed OS, to make this systems ready to compile and test applications.

## <a name="how-to-use">How to Use

[![asciicast](https://asciinema.org/a/366508.svg)](https://asciinema.org/a/366508)

#### Requirements

* Fedora (Tested on 22 to 32)
* DNF
* User whit SUDO credentials
* Perl
* perl-Switch
* perl-Time-Progress

**Fedora: Install packages**

```
sudo dnf install perl perl-switch perl-time-progress
```

## Usage

### Application configuration

See example file fedora-install-script.config

Separate the columns with TAB (tab-delimited)

This 2 parameters are the version of the configuration file:

```bash
# Reserved values:
ConfVersion	0.4
```

Configuration parameter documentation:

```
CFG Values:
  GenerateLog: Enable logs (true or false)
  LogPath: Directory to store the logfile
  LogMask: Mask for logging file
  Update: Update OS before install packages, this execute dnf update before process package input file (true or false)
  InstallBasedOnHostName: true or false
     NOTE: If this value is false, the script read the imput packages to install from the file general.config
     Otherwise the script read the imput packages from the file whit the current hostname in the filename
     Example: if the hostname command return this:
     $ hostname
     crt72339xx
     Then the imput file is crt72339xx.config
  RunSudo: run sudo command? (true or false)
  ShowProgressWindows: Show progress Message, otherwise only run in terminal. (true or false)
```

General Configuration:

Separate the columns with TAB (tab-delimited)

```bash
# GenerateLog	LogPath	LogMask	Update	InstallBasedOnHostName	RunSudo	ShowProgressWindows
true	$HOME/fedora-install.log	0600	true	true	true	true
```

Package selection configuration (the software to install in the OS)

See examples files general.config

general.config is the input file when the parameter InstallBasedOnHostName is false on the configuration file.

This 2 parameters are the version of the PackageSelection file:

```bash
# Reserved values:
PackageSelectionVersion	0.4
```

Configuration parameter documentation:

Values: <br />
  PackageName: Name of package <br />
  Category: Optional Category (see documentation) <br />

Configuration:

```bash
# PackageName	Category
screen	utility
mc	utility2
```

### Running the script

Target hostname:

If you run the script on the computer whit hostname NORC and the parameter InstallBasedOnHostName true in the configuration file, the file norc.config is used like input list for install the packages in the OS.

```bash
/PATH-TO-SCRIPT/fedora-install-script.pl --install_pkg
```

General package selection:

If you run the script on any computer whit the parameter InstallBasedOnHostName false, the file general.config is used like input list for install the packages in the OS.

```bash
/PATH-TO-SCRIPT/fedora-install-script.pl --install_pkg
```

Target hostname and filter install list by category:

If you run the script on the computer whit hostname NORC and the parameter InstallBasedOnHostName true in the configuration file, the file NORC.config is used like input list for install the packages in the OS but only the packages that matches whit the category selected are installed.

```bash
/PATH-TO-SCRIPT/fedora-install-script.pl --install_pkg_cat utility2
```

NOTE: Based on the configuration example on this documentation, only the package mc is installed.

General package selection  and filter install list by category:

If you run the script on any computer whit the parameter InstallBasedOnHostName false, the file general.config is used like input list for install the packages in the OS but only the packages that matches whit the category selected are installed

```bash
/PATH-TO-SCRIPT/fedora-install-script.pl --install_pkg_cat utility
```

NOTE: Based on the configuration example on this documentation, only the package terminator is installed.

## <a name="support-me">Support me:

### Librepay

<a href="https://liberapay.com/pablomenino/donate"><img alt="Donate using Liberapay" src="https://liberapay.com/assets/widgets/donate.svg"></a>

### Paypal

[![paypal](https://www.paypalobjects.com/en_US/i/btn/btn_donateCC_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=4HPTG85J8NQVG)

## <a name="third-party">Third party:

* **Icons:** The icon images are from [Tela Icon Theme](https://github.com/vinceliuice/Tela-icon-theme)


ChangeLog
Change Log:

    * 0.4.3 (2020/10/24)
      + Documentation

    * 0.4.2 (2020/10/20)
      + Documentation

    * 0.4.1 (2020-10-17)
      + Documentation

    * 0.4 (2019-02-10)
      + Clean up
      + Fix's

    * 0.3 (26-02-2016)
      + Enable logging
      + Fix's

    * 0.2 (11-02-2016)
      + Install category
      + Categorys  configurations
      + Configuration files
      + Docs

    * 0.1 (28-01-2016)
      + First candidate to test
      + Basic functions


test.pl
use Term::ReadKey;
my ($cols,$rows) = GetTerminalSize();

print "$cols,$rows";


fedora-install-script.pl
#!/usr/bin/perl

#----------------------------------------------------------------------------------------
# Fedora Install Script
# Version: 0.4.3
# 
# WebSite:
# https://lab.mfwlab.com/lab/fedora-install-script/
# 
# Copyright © 2020 - Pablo Meniño <pablo.menino@mfwlab.com>
#----------------------------------------------------------------------------------------

#----------------------------------------------------------------------
# Use declaration -----------------------------------------------------

use strict;
use warnings;
use Time::Local;
use Cwd;
use Cwd qw(abs_path);
use File::Basename;
use Switch;
use File::Path;
use Fcntl;
use Time::Progress;
use Term::ReadKey;

#----------------------------------------------------------------------
# Variables -----------------------------------------------------------

# Version Control
my $version = "0.4.2";
my $config_version = "0.4";
my $package_version = "0.4";

# Configuration file format ... that can be opened
my @version_check = ("0.1", "0.2", "0.3", "0.4");

# Package file format ... that can be opened
my @package_check = ("0.1", "0.2", "0.3", "0.4");

# Home directory
my $home = $ENV{"HOME"};

# Get Install Directory
my $script_dir = dirname(abs_path($0));

# Creates an unusual filename based on nanoseconds so that
# you don't accidentally overwrite another logfile.
my $nano = `date '+%d-%m-%Y_%H-%M-%S'`;
# Remove return line
chomp($nano);
$nano = "_" . $nano;

# Configuration variables
my ($GenerateLog, $LogPath, $LogMask, $Update, $InstallBasedOnHostName, $RunSudo, $ShowProgressWindows) = "";

# Configuration dir
my $cfg_dir_filename = $script_dir . "/";

# Configuration File
my $cfg_filename = $cfg_dir_filename . "fedora-install-script.config" ;

# Package variables
my ($PackageName, $CatName) = "";

# Command to execute
my $command = "" ;

# Get hostanme
my $host_name = `hostname`;
# Remove return line
chomp($host_name);

# Package list dir
my $pkg_dir_filename = $script_dir . "/";

# The location of executable
my $sudo = `which sudo`;
my $dnf = `which dnf`;
# Remove return line
chomp($sudo);
chomp($dnf);

# Get terminal size
my ($cols,$rows) = GetTerminalSize();

#----------------------------------------------------------------------
# Functions - print_help --------------------------------------------------------

sub print_help()
{
	print_version();
	print "Usage: $0 [options]\n";
	print "\n";
	print "options:\n";
	print "  --print_help                       - Print this help\n";
	print "  --print_version                    - Print version info\n";
	print "  --install_pkg                      - Install all packages from file\n";
	print "  --install_pkg_cat 'category_name'  - Install only the packages in the selected category\n";
	print "\n";
}

#----------------------------------------------------------------------
# Functions - print_version --------------------------------------------------------

sub print_version()
{
	print "Fedora Install script - Version $version\n";
	print "Copyright © 2020 - Pablo Meniño <pablo.menino\@mfwlab.com>\n";
	print "\n";
}

#----------------------------------------------------------------------
# Functions - read_config --------------------------------------------------------

sub read_config()
{
	# Return values:
	# 0: OK
	# 1: Not a valid config file
	# 2: Config File version not supported
	# 3: Executed failed

	my $nth = "";
	my $is_cfg_file = "false";
	my $version_cfg_file = "";
	my $return_value = 3;

	if (! -e $cfg_filename)
	{
		die "ERROR: File $cfg_filename doesn\'t exist\n";
	}

	# CFG
	open (FILECFG, $cfg_filename);
	while (<FILECFG>)
	{
		chomp($_);
		$nth = substr($_, 0, 1);
	
		# If not caracter NULL or #, then read config.
		if ( $nth ne "#" and $nth ne "" )
		{

			($GenerateLog, $LogPath, $LogMask, $Update, $InstallBasedOnHostName, $RunSudo, $ShowProgressWindows) = split( "\t", $_, 7);
		
			if ( $GenerateLog eq "ConfVersion" )
			{
				$is_cfg_file = "true";
				$version_cfg_file = $LogPath;
			}

		}

	}
	close (FILECFG);

	# If value version_check is found in cfg file ... then check if this is supported.
	if ($is_cfg_file eq "true" )
	{
		# Not supported version by default
		my $cfg_file_support = 1;
		my $version_check = "";
		foreach $version_check (@version_check)
		{
			if ( ($cfg_file_support == 1) && ($version_check eq $version_cfg_file) )
			{
				$cfg_file_support = 0;
			}
		}
		
		# If old version, then migrato to new.
		if ( $cfg_file_support == 1 )
		{
			# Not supported version
			$return_value = 2;
		}
		elsif ( $cfg_file_support == 0 )
		{
			# All checks are ok, continue
			$return_value = 0;
		}
		
	}
	else
	{
		# Not a valid config file
		$return_value = 1;
	}

	return $return_value;
}

#----------------------------------------------------------------------
# Functions - read_config_and_check --------------------------------------------------------

sub read_config_and_check()
{
	my $exit_code = read_config;

	if ( $exit_code == 1 )
	{
		die "ERROR: Not a valid config file\n";
	}
	elsif ( $exit_code == 2 )
	{
		die "ERROR: Config File version not supported\n";
	}
	elsif ( $exit_code == 3 )
	{
		die "ERROR: Executed failed\n";
	}
}

#----------------------------------------------------------------------
# Functions - read_package --------------------------------------------------------

sub read_package()
{
	
	my $category_pkg = $_[0];
	my $pkg_list_file = "";
	# Return values:
	# 0: OK
	# 1: Not a valid package file
	# 2: Package File version not supported
	# 3: Executed failed

	my $nth = "";
	my $is_cfg_file = "false";
	my $version_cfg_file = "";
	my $return_value = 3;

	# If InstallBasedOnHostName is true in config then load the file for that host
	if ( $InstallBasedOnHostName eq "true" )
	{
		$pkg_list_file = $pkg_dir_filename . $host_name . ".config";
	}
	else
	{
		$pkg_list_file = $pkg_dir_filename . "general.config";
	}

	if (! -e $pkg_list_file)
	{
		die "ERROR: File $pkg_list_file doesn\'t exist\n";
	}

	# PackageFile
	my @pkg_to_install;
	open (FILEPKG, $pkg_list_file);
	while (<FILEPKG>)
	{
		chomp($_);
		$nth = substr($_, 0, 1);
	
		# If not caracter NULL or #, then read config.
		if ( $nth ne "#" and $nth ne "" )
		{

			($PackageName, $CatName) = split( "\t", $_, 2);
		
			if ( $PackageName eq "PackageSelectionVersion" )
			{
				$is_cfg_file = "true";
				$version_cfg_file = $CatName;
			}
			elsif ( $category_pkg eq "default_all" )
			{
				push @pkg_to_install, $PackageName;
			}
			elsif ( $CatName eq $category_pkg )
			{
				push @pkg_to_install, $PackageName;
			}
		}

	}
	close (FILEPKG);
		
	# If value package_check is found in file ... then check if this is supported.
	if ($is_cfg_file eq "true" )
	{
		# Not supported package by default
		my $cfg_file_support = 1;
		my $version_check = "";
		foreach $version_check (@version_check)
		{
			if ( ($cfg_file_support == 1) && ($version_check eq $version_cfg_file) )
			{
				$cfg_file_support = 0;
			}
		}
		
		# If old version, then migrato to new.
		if ( $cfg_file_support == 1 )
		{
			# Not supported version
			$return_value = 2;
		}
		elsif ( $cfg_file_support == 0 )
		{
			# All checks are ok, continue
			# Install packages - Begin
			my $pkg_name;
			my $cmd;
			my $cmd_result = 0;
			my $cmd_result_0 = 0;
			my $cmd_result_other = 0;
			my $pkg_name_fix = "";
			
			my $pkg_number = 0;
			my $pkg_size = scalar @pkg_to_install;
			
			if ($pkg_size < 1)
			{
				die "ERROR: No packages to install. Check configuration files.\n";
			}

			# Check config for Update parameter, if this is true, run dnf update -y
			if ($Update eq "true")
			{
				print "Running DNF update\n";
	
				$cmd_result = update_cmd();
				if ($cmd_result != 0 )
				{
					die "ERROR: DNF update failed, check DNF logs for more information.\n";
				}
				else
				{
					print "\nDNF update done.\n\n";
				}

			}
			
			print "Install packages: The selected category is $category_pkg\n";
			
			$| = 1;
			my $p = new Time::Progress;

			$p->attr( min => 0, max => $pkg_size );

			# Fix size for progress bar
			my $cols_section1 = 0;
			if ($cols > 51)
			{
				$cols_section1 = int($cols-50);
			}
			
			foreach $pkg_name (@pkg_to_install)
			{
				
				$pkg_number=$pkg_number+1;

				my $sudo_tmp="";
				if ( $RunSudo eq "true" )
				{
					$sudo_tmp=$sudo;
				}
				else
				{
					$sudo_tmp="";
				}

				$cmd = $sudo_tmp . " " . $dnf. " -y install " . $pkg_name . " >/dev/null 2>&1";
												
				# Terminal size - cut string if more than 20 caracters
				$pkg_name_fix = pack("A20",$pkg_name);
			
				# If terminal size is lower than 51, auto adjust progress bar size
				if ($cols_section1 == 0)
				{
					print $p->report("[" . $pkg_number . "/" . $pkg_size ."] $pkg_name_fix |" . "%L %p\r", $pkg_number);
				}
				else
				{
					print $p->report("[" . $pkg_number . "/" . $pkg_size ."] $pkg_name_fix |" . "%L %". $cols_section1 ."b %p\r", $pkg_number);
				}

				$cmd_result = &execute_cmd( $cmd );
				if ($cmd_result == 0 )
				{
					$cmd_result_0=$cmd_result_0+1;
				}
				else
				{
					$cmd_result_other=$cmd_result_other+1;
				}
			}
			# Install packages - End
			print $p->report("done %p elapsed: %L (%l sec)", $pkg_number);
			print "\n\nInstalled packages:\n";
			print "Successful: " . $cmd_result_0 . " | Failed: " . $cmd_result_other . "\n";
			
			if ($cmd_result_other > 0)
			{
				print "\nWARNING: They are $cmd_result_other packages failed, see log file for more information.\n";
			}
			
			$return_value = 0;

		}
		
	}
	else
	{
		# Not a valid package file
		$return_value = 1;
	}

	return $return_value;
}

#----------------------------------------------------------------------
# Functions - read_package_and_check --------------------------------------------------------

sub read_package_and_check()
{
	my $category_pkg = $_[0];
	my $exit_code = &read_package($category_pkg);

	if ( $exit_code == 1 )
	{
		die "ERROR: Not a valid package file\n";
	}
	elsif ( $exit_code == 2 )
	{
		die "ERROR: Package File version not supported\n";
	}
	elsif ( $exit_code == 3 )
	{
		die "ERROR: Executed failed\n";
	}
}

#----------------------------------------------------------------------
# Functions - execute_cmd --------------------------------------------------------

sub execute_cmd()
{
	my $systemCommand = $_[0];
    my $returnCode = system( $systemCommand );
    return $returnCode;
}

#----------------------------------------------------------------------
# Functions - update_cmd --------------------------------------------------------

sub update_cmd()
{
	my $sudo_tmp="";

	if ( $RunSudo eq "true" )
	{
		$sudo_tmp=$sudo;
	}
	else
	{
		$sudo_tmp="";
	}

	my $returnCode = &execute_cmd($sudo_tmp . " " . $dnf. " -y update");
	return $returnCode;
}

#----------------------------------------------------------------------
# Functions - install_pkg_cat --------------------------------------------------------

sub install_pkg_cat()
{
	my $category_pkg = $_[0];
	# Install Packages
	&read_package_and_check($category_pkg);
}

#----------------------------------------------------------------------
# Functions - check_software --------------------------------------------------------

sub check_software()
{
	if ( $sudo eq "" or $dnf eq "" )
	{
		die "ERROR: DNF or SUDO binary files not found.\n";
	}
}

#----------------------------------------------------------------------
# Main - Begin --------------------------------------------------------

# Check script arguments.
if ($#ARGV < 0)
{
	print_help();
}
else
{
	switch ($ARGV[0])
	{
		case "--print_help"
		{
			print_help();
		}
		case "--print_version"
		{
			print_version();
		}
		case "--install_pkg"
		{
			# Print version
			print_version();
			# Check SUDO and DNF
			check_software();
			# Read Config
			read_config_and_check();
			# Install
			&install_pkg_cat( "default_all" );
		}
		case "--install_pkg_cat"
		{
			# Print version
			print_version();
			# Check SUDO and DNF
			check_software();
			# Check category input string
			if ( $#ARGV >= 1 )
			{
				# Read Config
				read_config_and_check();
				# Install
				&install_pkg_cat($ARGV[1]);
			}
			else
			{
				die "ERROR: No category selected\n";
			}
		}
		else
		{
			print_help();
		}
	}
}

#----------------------------------------------------------------------
# Main - End ----------------------------------------------------------
#----------------------------------------------------------------------
